### DISCUSSION TITLE
Make habit list horizontally scrollable

### DISCUSSION THREAD

**Discussion by chennemann** [iSoron/uhabits#466] [Feature Requests]
I don't know whether this was already suggested but I think it could be a nice thing to make the habit list horizontally scrollable so that you are able to see more than just the last five days. 

Maybe the scrolling should be a little bit harder to trigger. What I mean by that is that it starts not simultaneously with the finger movement but with a bit of a threshold.

**Comment by iSoron** [comment]
@chennemann You can already scroll the list horizontally by dragging the header (the grey part on the top which shows the days of the week). I agree that this is not very discoverable. Please feel free to build a prototype where the list can be dragged horizontally anywhere. As you mentioned, getting this right might be a little complicated.

**Comment by chennemann** [comment]
Awesome. I didn't know that. The discoverability of my solution would not be better I guess. 

I think we should just add another "did you know" to promote this feature.

It should show up as soon as the user has one habit checked more than the visible days ago. 

What do you think about that?

**Comment by ImAnnoying2** [comment]
Maybe an option to launch the app in landscape by default could be a quick solution for those that have fewer habits but want to get motivated by streaks. Or maybe list habit streaks vertically instead of horizontally, also as an option?

**Comment by tonypuglieso** [comment]
Hello other posible solutions, Add a symbol so that the user knows that it can be moved and / or a smaller mode, so that in the vertical mode more days are filled.

![loop habits sketch](https://user-images.githubusercontent.com/52177646/83360864-f1f57180-a35a-11ea-9837-2bdbfb7d8f51.png)

I understand that a minimal mode, it can be difficult to press such small buttons. That could be a problem. But it would be an interesting option that someone can manually activate in the settings.

Other posible symbols example.

![loop habits sketch2](https://user-images.githubusercontent.com/52177646/83361130-b9569780-a35c-11ea-96a8-57945de3ab22.png)

**Comment by NomarCub** [comment]
Scrolling into the future in general is a feature I would kill for. Is this issue now more for the scrolling UI and FAQ adjustments, or in general scrolling forward in time? If it's the former, maybe the latter should have it's own discussion. Planning ahead also came up in other issues - https://github.com/iSoron/uhabits/discussions/372.

**Reply by NomarCub** [reply]
I tried my hand with the code, but the current day looks kind of baked into how the habit list is displayed now. Is there someone who could tell me how to go about this?
@iSoron, should I open an issue? I see that feature requests are supposedly moved here, but discussions seem pretty inactive, and issues are still going on. Or should I just open a new discussion specific to this, as previously suggested?

**Reply by hiqua** [reply]
If I were you I'd grep the code for `today` and see where the first day is determined, try to have that as an argument you can pass instead of something determined by the clock, and adapt some of the logic that does the scrolling to the right for previous days.

Please note that one possibly relevant class is to be deprecated: https://github.com/iSoron/uhabits/issues/1051. Doing this issue could actually also help to understand the code (but no guarantee).

Feel free to join https://gitter.im/loophabits/dev to ask more questions, I probably won't be able to answer all of them, but I can at least try.

**Reply by iSoron** [reply]
@NomarCub I expect this feature to be hard to implement, since the app was not originally designed to compute or display checkmarks in the future. I think the first step would be ensuring that the app computes values for the future. We can then focus on visualizing them. In other words, you would need to modify the `org.isoron.uhabits.core.models` package. Please feel free to open a separate discussion thread in the "Technical Discussion" category or to submit a draft pull request.

@hiqua You can find which classes are querying the current date by finding usages of `DateUtils.getToday` and `DateUtils.getTodayWithOffset`. It's possible to change "today" globally by calling `DateUtils.setFixedLocalTime` but this cannot be used to implement this feature, because sometimes you do need the correct current date, such as when scheduling reminders.

**Reply by NomarCub** [reply]
I've been looking for a quick and dirty local solution, and found this:
```kt
if (prefs.isMidnightDelayEnabled) {
    setStartDayOffset(3, 0)
}
```

Then set 3 to a negative value. It's not meant for this, but works with app restarts, with a fixed value. Could this be a basis for a proper solution?

**Reply by NomarCub** [reply]
For anyone interested, here's the modification: https://github.com/NomarCub/uhabits/pull/1
You can get the modified apk from GitHub also: https://github.com/NomarCub/uhabits/actions/runs/3260492784
